File: Leetcode/14.Longest_Common_Prefix/14.Longest_Common_Prefix.js
const longestCommonPrefix = (strs) => {
  const zip = (...strs) =>
    [...strs[0]].map((_, idx) => strs.map((str) => str[idx]));

  if (!strs.length) return "";
  const zipArray = zip(...strs);
  let prefix = "";
  for (let i = 0; i < zipArray.length; i++) {
    const setObj = new Set(zipArray[i]);
    const isSame = setObj.size === 1;
    if (isSame) {
      prefix += zipArray[i][0];
    } else {
      return prefix;
    }
  }
  return prefix;
};

const strs = ["flower", "flow", "flight"];
console.log(longestCommonPrefix(strs));

File: Leetcode/14.Longest_Common_Prefix/14.Longest_Common_Prefix.py
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """

        result = []
        for x in zip(*strs):
            if len(set(x)) == 1:
                result.append(x[0])
            else:
                break

        return "".join(result)


strs = ["flower", "flow", "flight"]
# strs = ["dog", "racecar", "car"]
# strs = ["ab", "a"]

sol = Solution()

print(sol.longestCommonPrefix(strs))
